keps/sig-node/2043-pod-resource-concrete-assigments/README.md

```diff
@@ -0,0 +1,402 @@
+the resources managed by the new `HintProvider`.
+ContainerDevices reflects information obtained from device plugins. For example, device plugin could describe
+resource which spread across NUMA nodes. Also it's possible to have a device located on the different NUMA nodes.
+ContainerDevices could represent it as following:
+```json
+[
+    {
+        resource_name: "res1",
+        device_ids: [ "dev1", "dev2" ],
+        topology: [ { ID: 0 }, ],
+    },
+    {
+        resource_name: "res1",
+        device_ids: [ "dev3", "dev4" ],
+        topology: [ { ID: 1 } ]
+    },
+    {
+        resource_name: "res2",
+        device_ids: [ "dev5" ],
+        topology: [ { ID: 1 }, { ID: 2 } ]
+    },
+]
+```
+
+Using the `Watch` endpoint, client applications can be notified of the pod resource allocation changes as soon as possible.
+However, the state of a pod will not be sent up until the first resource allocation change, which is the pod deletion in the worst case.
+Client applications who need to have the complete resource allocation picture thus need to consume both `List` and `Watch` endpoints.
+
+The `resourceVersion` found in the responses of both APIs allows client applications to identify the most recent information.
+The `resourceVersion` value is updated following the same semantics of pod `resourceVersion` value, and the implementation
+may use the same value from the corresponding pods.
+To keep the implementation simple as possible, the kubelet does *not* store any historical list of changes.
+
+In order to make sure not to miss any updates, client application can:
+1. call the `Watch` endpoint to get a stream of changes.
+2. call the `List` endpoint to get the state of all the pods in the node.
+3. reconcile updates using the `resourceVersion`.
+
+In order to make the resource accounting on the client side, safe and easy as possible the `Watch` implementation
+will guarantee ordering of the event delivery in such a way that the capacity invariants are always preserved, and the value
+will be consistent after each event received - not only at steady state.
+Consider the following scenario with 10 devices, all allocated: pod A with device D1 allocated gets deleted, then
+pod B starts and gets device D1 again. In this case `Watch` will guarantee that `DELETE` and `ADDED` events are delivered
+in the correct order.
+
+### Test Plan
+
+Given that the API allows observing what device has been associated to what container, we need to be testing different configurations, such as:
+* Pods without devices assigned to any containers.
+* Pods with devices assigned to some but not all containers.
+- Pods with devices assigned to init containers.
+- ...
+
+We have identified two main ways of testing this API:
+- Unit Tests which won't rely on gRPC. They will test different configurations of pods and devices.
+- Node E2E tests which will allow us to test the service itself.
```

Comment:
I'd like to explicitly call out testing for both a list-only client, and a list-and-watch client.

Reply:
very fair point. My implementation plan is to add first e2e test to capture the existing behaviour and then start making changes. See POC here: https://github.com/kubernetes/kubernetes/pull/94612/commits
[this commit](https://github.com/kubernetes/kubernetes/pull/94612/commits/a9184e1811f34c6546c869183a2cfe2ddd5b2262) can (and should) be put at the beginning of the series. I can even extract it in a separate PR if it's preferred.

Reply:
no, that works for me.  I'm just asking for it to be documented in the KEP.

Reply:
Sure, will post an update addressing this comment.